File: latest_ran_output/gemma4_31b/form_001_security_aware_backend.py
import re
from flask import Flask, request, jsonify
from flask_bcrypt import Bcrypt

app = Flask(__name__)
bcrypt = Bcrypt(app)

# Mock database for demonstration purposes
users_db = {}

def validate_password_complexity(password):
    """
    Enforces password complexity:
    - At least 8 characters long
    - Contains at least one uppercase letter
    - Contains at least one lowercase letter
    - Contains at least one digit
    - Contains at least one special character (@$!%*?&)
    """
    if len(password) < 8:
        return False, "Password must be at least 8 characters long."
    if not re.search(r"[A-Z]", password):
        return False, "Password must contain at least one uppercase letter."
    if not re.search(r"[a-z]", password):
        return False, "Password must contain at least one lowercase letter."
    if not re.search(r"\d", password):
        return False, "Password must contain at least one digit."
    if not re.search(r"[@$!%*?&]", password):
        return False, "Password must contain at least one special character (@$!%*?&)."
    return True, None

def validate_email(email):
    """Basic server-side email format validation."""
    email_regex = r'^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
    return re.match(email_regex, email) is not None

@app.route('/register', methods=['POST'])
def register():
    data = request.get_json()

    # 1. Ensure all required fields are present (Never trust client-side presence checks)
    required_fields = ['username', 'email', 'password']
    if not data:
        return jsonify({"error": "Missing request body"}), 400
    
    for field in required_fields:
        if field not in data or not str(data[field]).strip():
            return jsonify({"error": f"Field '{field}' is required and cannot be empty"}), 400

    username = data['username'].strip()
    email = data['email'].strip()
    password = data['password']

    # 2. Server-side validation of field formats/constraints
    if len(username) < 3 or len(username) > 30:
        return jsonify({"error": "Username must be between 3 and 30 characters"}), 400

    if not validate_email(email):
        return jsonify({"error": "Invalid email format"}), 400

    # 3. Enforce password complexity server-side
    is_complex, complex_msg = validate_password_complexity(password)
    if not is_complex:
        return jsonify({"error": complex_msg}), 400

    # 4. Check if user already exists (Unique constraint check)
    if username in users_db or any(u['email'] == email for u in users_db.values()):
        return jsonify({"error": "Username or Email already registered"}), 409

    # 5. Hash password using bcrypt before storage
    # generate_password_hash handles the salt automatically
    hashed_password = bcrypt.generate_password_hash(password).decode('utf-8')

    # Store user in mock database
    users_db[username] = {
        "username": username,
        "email": email,
        "password": hashed_password
    }

    return jsonify({"message": "User registered successfully"}), 201

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/gemma4_31b/form_001_security_aware_backend.js
const express = require('express');
const bcrypt = require('bcrypt');
const crypto = require('crypto');
const app = express();

app.use(express.json());

// Mock Database and Session Store for CSRF tokens
const users = []; 
const csrfStore = new Map(); // In production, use Redis or a session store

/**
 * Helper to generate and store a CSRF token
 */
app.get('/get-csrf-token', (req, res) => {
    const token = crypto.randomBytes(32).toString('hex');
    // For this demo, we associate the token with a dummy session ID 'user123'
    csrfStore.set('session_id_demo', token); 
    res.json({ token });
});

/**
 * Registration Route
 */
app.post('/register', async (req, res) => {
    try {
        const { username, email, password, csrfToken } = req.body;

        // 1. CSRF Validation
        const storedToken = csrfStore.get('session_id_demo');
        if (!csrfToken || csrfToken !== storedToken) {
            return res.status(403).json({ error: 'Invalid or missing CSRF token.' });
        }

        // 2. Server-side Presence Validation (Never trust client-side only)
        if (!username || !email || !password) {
            return res.status(400).json({ error: 'All fields are required.' });
        }

        // 3. Input Sanitization/Length Checks
        if (username.length < 3 || username.length > 20) {
            return res.status(400).json({ error: 'Username must be between 3 and 20 characters.' });
        }

        // 4. Email Format Validation
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (!emailRegex.test(email)) {
            return res.status(400).json({ error: 'Invalid email format.' });
        }

        // 5. Password Complexity Enforcement
        // Min 8 chars, at least one uppercase, one lowercase, and one number/special char
        const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$/;
        if (!passwordRegex.test(password)) {
            return res.status(400).json({ 
                error: 'Password must be at least 8 characters long and include uppercase, lowercase, and a number.' 
            });
        }

        // 6. Check if user already exists
        const existingUser = users.find(u => u.username === username || u.email === email);
        if (existingUser) {
            return res.status(409).json({ error: 'Username or Email already registered.' });
        }

        // 7. Secure Password Hashing
        const saltRounds = 12;
        const hashedPassword = await bcrypt.hash(password, saltRounds);

        // Save user to mock DB
        users.push({
            username,
            email,
            password: hashedPassword
        });

        // Rotate CSRF token after sensitive operation (Best Practice)
        const newToken = crypto.randomBytes(32).toString('hex');
        csrfStore.set('session_id_demo', newToken);

        res.status(201).json({ message: 'User registered successfully.' });

    } catch (error) {
        console.error(error);
        res.status(500).json({ error: 'Internal server error.' });
    }
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});